src/fabric/annealing.rs

```diff
@@ -94,56 +172,134 @@ impl MeasurePair {
 
 struct PairGenerator {
     joints: Vec<JointIncident>,
-    pairs: HashMap<(usize, usize), MeasurePair>,
+    intervals: HashMap<(usize, usize), Interval>,
+    pairs: HashMap<(usize, usize), Pair>,
 }
 
 impl PairGenerator {
-    fn new(joints: Vec<JointIncident>) -> Self {
+    fn new(joints: Vec<JointIncident>, intervals: HashMap<(usize, usize), Interval>) -> Self {
         Self {
             joints,
+            intervals,
             pairs: HashMap::new(),
         }
     }
 
-    fn generate_pairs(mut self) -> impl Iterator<Item=MeasurePair> {
-        for joint in self.joints.clone() {
-            self.add_pairs_for(joint)
+    fn proximity_measures(mut self) -> impl Iterator<Item=Pair> {
+        for joint in 0..self.joints.len() {
+            self.push_proximity(joint);
         }
```

Comment:
Hm. Are you passing an index instead of a JointIncident to avoid the clone()? If so, couldn't you pass `&JointIncident`s instead? This seems less Rusty.

Reply:
that causes a borrowing problem